lua/swampshop/sh_products.lua

```diff
@@ -83,28 +83,52 @@ function SS_WeaponProduct(product)
         self:SS_WeaponProduct_OnBuy(ply)
     end
 
+    function product:CannotBuy(ply)
+        return ply:HasWeapon(self.Class) and "You're already carrying this."
+    end
+
     SS_DeathKeepnotice(product)
     SS_Product(product)
 end
 
 function SS_WeaponAndAmmoProduct(product)
     function product:OnBuy(ply)
         if not ply:HasWeapon(self.class) then
-            ply:Give(self.class)
+            local wep = ply:Give(self.class)
+            
+            if(self.amount)then --ensure the player has at least this many ammo when giving a new one to them. Otherwise, give them whatever is standard.
+                local ammotype = self.ammotype or game.GetAmmoName(wep:GetPrimaryAmmoType())
+                local curammo = ply:GetAmmoCount(self.ammotype)
+                ply:SetAmmo(math.max(curammo,self.amount),ammotype)
+            end
+        else
             local wep = ply:GetWeapon(self.class)
+            local ammotype = self.ammotype or game.GetAmmoName(wep:GetPrimaryAmmoType())
+            local ammogive = self.amount or (wep.Primary and wep.Primary.DefaultClip) or wep:GetMaxClip1() or 0
 
-            if wep:Clip1() > 0 then
-                wep:SetClip1(0)
+            if (ammogive <= 0) then
+                ammogive = nil
             end
 
-            ply:SetAmmo(self.amount, self.ammotype)
-        else
-            ply:GiveAmmo(self.amount, self.ammotype)
+            if (ammotype and ammogive) then
+                ply:GiveAmmo(ammogive, ammotype)
+            end
         end
-
         ply:SelectWeapon(self.class)
     end
-
+    function product:CannotBuy(ply)
+        if (ply:HasWeapon(self.class)) then
+            local wep = ply:GetWeapon(self.class)
+            local ammotype = self.ammotype or (game.GetAmmoName(wep:GetPrimaryAmmoType()))
+            local ammogive = self.amount or (wep.Primary and wep.Primary.DefaultClip) or wep:GetMaxClip1() or 0
```

Comment:
Does :GetMaxClip1() return the same thing as (wep.Primary and wep.Primary.DefaultClip) for sweps?

Reply:
I'm honestly not sure, I got inconsistent results, but defaultclip is more reliable, since it's a user editable var